### ISSUE TITLE
WS-Federation - add claims to request 

### ISSUE THREAD

**Issue by michalor** [dotnet/wcf#4979]
Hello. 
I am currently building a connector to WS-Federation service and I am trying to issue token from identity provider. In my use case I need to pass some claims to it as a part of the RequestSecurityToken message. I just wonder if there is a way to do so in current dotnet/wcf implementation. I can see a possibility to add Claims to WSTrustTokenParameters, but they are not getting into the final request. 

This is my current code:

```
public async Task<System.IdentityModel.Tokens.SecurityToken> GetTokenFromIPAsRegistrationAttemptUserName(string urlIP, string appliesTo, string username, string password, string controlQuestionID, string controlQuestionAnswer, string captchaResponseHash, Claims claims)
{
    var ipAddr = new EndpointAddress(urlIP + "/issue.svc/username");
    var binding = new WS2007HttpBinding();
    binding.Security.Mode = SecurityMode.TransportWithMessageCredential;
    binding.Security.Transport.ClientCredentialType = HttpClientCredentialType.None;
    binding.Security.Message.ClientCredentialType = MessageCredentialType.UserName;
    binding.Security.Message.EstablishSecurityContext = false;
    binding.Security.Message.NegotiateServiceCredential= false;
    binding.UseDefaultWebProxy = true;
    binding.Security.Message.AlgorithmSuite = System.ServiceModel.Security.SecurityAlgorithmSuite.Default;
    
    var tokenParams = WSTrustTokenParameters.CreateWS2007FederationTokenParameters(binding, ipAddr);

    tokenParams.KeyType = System.IdentityModel.Tokens.SecurityKeyType.SymmetricKey; 
    tokenParams.ClaimTypes.Add(claims);

    var clientCredentials = new ClientCredentials();
    clientCredentials.UserName.UserName = username;
    clientCredentials.UserName.Password = password;
    var trustCredentials = new WSTrustChannelClientCredentials(clientCredentials);
    var tokenManager = trustCredentials.CreateSecurityTokenManager();
    var tokenRequirement = new System.IdentityModel.Selectors.SecurityTokenRequirement();
    tokenRequirement.TokenType = "urn:oasis:names:tc:SAML:1.0:assertion";

tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/IssuedSecurityTokenParameters"] = tokenParams;
    tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/TargetAddress"] = new EndpointAddress(appliesTo + "/issue.svc");
tokenRequirement.Properties["http://schemas.microsoft.com/ws/2006/05/servicemodel/securitytokenrequirement/SecurityAlgorithmSuite"] = System.ServiceModel.Security.SecurityAlgorithmSuite.Default;
   
    var tokenProvider = tokenManager.CreateSecurityTokenProvider(tokenRequirement);

    ((ICommunicationObject)tokenProvider).Open();
    var token = await tokenProvider.GetTokenAsync(Timeout.InfiniteTimeSpan);
    ((ICommunicationObject)tokenProvider).Close();
    return token;

}
```

Could you please help me understand if I am missing something or if this is not implemented?

**Comment by HongGit** [comment]
@mconnew can you please comment?

@imcarolwang can you please look into the implementation and add a unit test for this?

**Comment by mconnew** [comment]
@michalor, this does look like a bug. We also don't copy across the AdditionalRequestParameters data either. I thought you might be able to work around this as the class `WSTrustChannelSecurityTokenProvider` is public, and the method `CreateWsTrustRequest` is virtual, but you will hit another problem which is `WSTrustChannelSecurityTokenProvider.ClientCredentials` is internal and there's no way to set it without using reflection. If you can tolerate setting `ClientCredentials` using reflection, then it's doable. If not, then we will need to ship a fix for you to be able to do what you are wanting.

**Comment by michalor** [comment]
Thank you for the suggestion! 👍  
I have implemented a derived `WSTrustChannelSecurityTokenProviderWithClaims` class like this it like this:

```
    public class WSTrustChannelSecurityTokenProviderWithClaims : WSTrustChannelSecurityTokenProvider
    {
        public Claims claims { get; set; }

        public WSTrustChannelSecurityTokenProviderWithClaims(SecurityTokenRequirement tokenRequirement) : base(tokenRequirement)
        { }

        protected override WsTrustRequest CreateWsTrustRequest()
        {
            var trustRequest = base.CreateWsTrustRequest();
            trustRequest.Claims = this.claims;
            
            return trustRequest;
        }
    }
```

And changed a few lines in my `GetTokenFromFP` method like this:

```
var tokenProvider = new WSTrustChannelSecurityTokenProviderWithClaims(tokenRequirement);
typeof(WSTrustChannelSecurityTokenProvider).GetProperty("ClientCredentials", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance).SetValue(tokenProvider, clientCredentials.Clone());
tokenProvider.claims = claims;
```

This approche works fine :-), so I am good for now and I am looking forward to proper implementation.

**Comment by mconnew** [comment]
I'm glad I helped you get a working solution. I'm going to open a new issue to track the code fixes which are required and close this one as you are unblocked. I'll include making ClientCredentials public as without that it makes no sense to have the class derivable.
